src/bridge/cockpitdisksamples.c

```diff
@@ -153,3 +158,130 @@ cockpit_disk_samples (CockpitSamples *samples)
   g_strfreev (lines);
   g_free (contents);
 }
+
+static gchar*
+read_file (const gchar *path)
+{
+  g_autoptr(GError) error = NULL;
+  gchar *contents = NULL;
+  if (!g_file_get_contents (path, &contents, NULL, &error))
+    {
+      /* Do not log:
+       * - file does not exist: process died and can no longer be sampled
+       * - access denied: bridge is not privileged and can't sample all processes
+       * - sometimes ERROR_FAILED is returned with message: "No such process"
+       */
+      if (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_NOENT) ||
+          g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_ACCES) ||
+          (g_error_matches (error, G_FILE_ERROR, G_FILE_ERROR_FAILED) && g_strrstr(error->message, "No such process")))
+        {
+          return NULL;
+        }
+
+      g_warning ("Domain: %s, code: %d, message: %s", g_quark_to_string (error->domain), error->code, error->message);
+      return NULL;
+    }
+
+  return contents;
+}
+
+static void
+table_add_values(GHashTable *table,
+                   gchar *cgroup,
+                   guint64 read,
+                   guint64 write)
+{
+  cgroup_values_t *values = g_hash_table_lookup (table, cgroup);
+  if (values == NULL)
+    {
+      values = g_new0(cgroup_values_t, 1);
+      g_hash_table_insert (table, g_strdup (cgroup), values);
+    }
+
+  values->read = read;
+  values->write = write;
+}
+
+static void
+get_process_io (const gchar *dir_name,
+                GHashTable *table)
+{
+  g_autofree gchar *path = g_strdup_printf ("/proc/%s/io", dir_name);
+  g_autofree gchar *contents = read_file (path);
+  if (!contents)
+    return;
+
+  guint64 read = 0, write = 0;
+  gchar **lines = g_strsplit (contents, "\n", -1);
+  for (guint n = 0; lines != NULL && lines[n] != NULL; n++)
+    {
+      gchar description[32] = {'\0'};
+      guint64 value;
+      const gchar *line = lines[n];
+      if (strlen (line) == 0)
+        continue;
+
+      if (sscanf (line, "%s %" G_GUINT64_FORMAT, description, &value) != 2)
+        {
+          g_warning ("Error parsing line %d of file /proc/%s/io: `%s'", n, dir_name, line);
+          continue;
+        }
+
+      if (g_str_equal (description, "read_bytes:"))
+        read = value;
+      else if (g_str_equal (description, "write_bytes:"))
+        write = value;
+    }
+
+  g_strfreev (lines);
+
+  // get process cgroup
+  g_autofree gchar *cgroup_path = g_strdup_printf ("/proc/%s/cgroup", dir_name);
+  g_autofree gchar *cgroup = read_file (cgroup_path);
```

Comment:
You could use an `fopen` wrapper here which implements:

 - dirfd opening + fdopen handling
 - all the error handling you want to do

and then, `FILE *` in hand, reading the cgroup is trivial.

Then you could modify this entire function to take:

1) dirfd to the process subdir
2) pointer to hash table